### DISCUSSION TITLE
How do I use 'onTransition' when I'm using 'useMachine'?

### DISCUSSION THREAD

**Discussion by felippenardi** [statelyai/xstate#1734] [General]
This is the way I'm starting my machine in React:

```js
  const [state, send] = useMachine(machine, { devTools: true })
```

How do I hook up a log? I found that apparently I should use the `onTransition` method, but I can't find where should I get it from.

**Comment by f-elix** [comment]
The service is returned as a 3rd value of the tupple. You can then use `onTransition` directly on it.

**Comment by davidkpiano** [comment]
1. `onTransition` is almost the same as just "reacting" to state changes:


```js
const [state, send] = useMachine(machine);

useEffect(() => {
  // called whenever the state changes
  console.log(state);
}, [state]);
```

2. You also have access to the service directly as the third item in the tuple:

```js
const [state, send, service] = useMachine(machine);

useEffect(() => {
  const handler = (state) => { console.log(state) }

  service.onTransition(handler);

  return () => { service.off(handler); }
}, []);
```

I'd strongly recommend option 1, as it is simpler.

**Reply by Andarist** [reply]
@felippenardi keep in mind that those 2 solutions are not 100% equivalent though because of how React `setState` batching works - the first approach might "miss" some states.

**Reply by BrunoQuaresma** [reply]
Is there a way to get the event during onTransition? Use case: Having a machine with two states, "started" and "idle", I would like to execute an outside action from the UI when "started" became "idle", something like a "onIdle" callback.
